File: 프로그래머스/Level2/큰 수 만들기.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

string solution(string number, int k) {
    string answer;
    stack<char> s; // 정답을 담을 stack.
    
    for(int i=0; i<number.size(); i++){
        while(!s.empty() && k > 0){
            if(s.top() < number[i]){
                s.pop();
                k--;
            }
            else
                break;
        }
        s.push(number[i]);
    }
    
    while(k > 0){
        s.pop();
        k--;
    }
    
    
    while(!s.empty()){
        answer = s.top() + answer;
        s.pop();
    }

    return answer;
}

File: 프로그래머스/Level2/큰 수 만들기.java
import java.util.*;

class Solution {
    public String solution(String number, int k) {
        
        int answerLen = number.length() - k;
        
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < number.length(); ++i) {
            int num = number.charAt(i) - '0';
            while (!stack.isEmpty() && stack.peek() < num && k > 0) {
                stack.pop();
                k--;
            }
            stack.push(num);
        }
        
        StringBuilder answer = new StringBuilder();
        for (int i = 0; i < answerLen; ++i) {
            answer.append(stack.get(i));
        }
        
        return answer.toString();
    }
}
